### DISCUSSION TITLE
[Feature] XMP Sidecar for Read-Only External Libraries

### DISCUSSION THREAD

**Discussion by flammermann** [immich-app/immich#12001] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I would propose to enable creation/modification of XMP sidecar files also for setups using an external library that is set to read-only (ro flag in docker mount).

Currently, some features require creation/modification of the XMP sidecar file. For example: Star rating and changing date, time, or location. I expect future features to also make use of this mechanism, for example cropping and rotation. However, with a read-only external library this silently fails (no error in UI, only in the logs of the server).

It would be great to also support these feature for read-only external libraries by:

1. Only storing the information in the database and skip the XMP sidecar creation/modification. Maybe this could be toggled as a setting per external library.
2. Creating the XMP sidecar in a separate writable mount mimicking the folder structure from the external library.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by RobLW** [comment]
How are you rating and changing dates on read-only external libraries?
Surely the point of it being read-only is that it doesn't run the risk of writing anything to the location of the system which these files are stored. Perhaps storing it in the database certainly, but this won't be reflected outside.

I think generally speaking if you are pulling in external libraries which are read-only, you do the editing in your own app away from Immich, and Immich are trying to build a process which picks up on the changes in that directory.

**Reply by bo0tzz** [reply]
This is indeed the idea. To address the points in the original post:
1. This would mean the data gets overwritten if you run metadata extraction for that asset again (with the data from the original file). 
2. Storing _sidecar_ files separately kind of defeats their point.

**Reply by JanGross** [reply]
> Storing sidecar files separately kind of defeats their point.

Fair point. However, this would at least allow for periodic manual merging of sidecar files into the original folder.  
I for example only mark my library as read only to prevent deletion. AFAIK there is no way to set a mount as write only.

> This would mean the data gets overwritten if you run metadata extraction for that asset again (with the data from the original file).

Maybe it'd be a solution to have a `'locked'` field with each metadata property or per file indicating that this piece of data has already been touched and shouldn't be overwritten unless specifically unlocked.

**Reply by ludditus-RPI** [reply]
I believe this issue prevents a user from tagging photos that are in an external library mounted as read-only.  I can tag photos that I upload but not photos in an external library on a read-only network share. The log entry is "error reading exif data (path/file.xmp): Error: File not found - /path/file.xmp.

Immich will let you go through the motions of tagging photos in a read-only drive, but the tag doesn't persist. It would be nice if there was a mechanism to allow tagging over photos in this situation.

**Reply by kvalev** [reply]
You can see this other discussion for some ideas how you can handle that on your end https://github.com/immich-app/immich/discussions/5530
